Linker error when using `__builtin_available()` version checks higher than `macosx-version-min`

Host: Fedora 34
OSXCross: 8a716a43a72dab1db9630d7824ee0af3730cb8f9
SDK: Xcode 12.4, MacOSX 11.1

This is a follow-up to https://github.com/FNA-XNA/FNA/issues/318, both FNA and Godot are building Mono with osxcross for x86_64 and arm64, and we run into issues when using a `macosx-version-min` lower than 11.0 while linking to Mono which uses a `__builtin_available(macOS 11, *)` version check here: https://github.com/mono/mono/blob/a01309d1041ff8ae15b1f8b717420ec139e1cdb8/mono/utils/mono-codeman.c#L670 (note this is a recent addition to Mono's `2020-02` branch, which is the current Preview release).

Here's an example of the error from building Godot with osxcross 8a716a43 and `mono-6.12.0.144`:
```
/root/osxcross/target/bin/x86_64-apple-darwin20.2-c++ -o bin/godot.osx.opt.tools.x86_64.mono -arch x86_64 -mmacosx-version-min=10.12 -framework Cocoa -framework Carbon -framework OpenGL -framework AGL -framework AudioUnit -framework CoreAudio -framework CoreMIDI -lz -framework IOKit -framework ForceFeedback -framework AVFoundation -framework CoreMedia -framework CoreVideo -rdynamic -Wl,-force_load,/root/dependencies/mono/lib/libmonosgen-2.0.a platform/osx/crash_handler_osx.osx.opt.tools.x86_64.o platform/osx/os_osx.osx.opt.tools.x86_64.o platform/osx/godot_main_osx.osx.opt.tools.x86_64.o platform/osx/dir_access_osx.osx.opt.tools.x86_64.o platform/osx/joypad_osx.osx.opt.tools.x86_64.o platform/osx/power_osx.osx.opt.tools.x86_64.o -L/root/dependencies/mono/lib main/libmain.osx.opt.tools.x86_64.a main/tests/libtests.osx.opt.tools.x86_64.a modules/libmodules.osx.opt.tools.x86_64.a platform/libplatform.osx.opt.tools.x86_64.a drivers/libdrivers.osx.opt.tools.x86_64.a editor/libeditor.osx.opt.tools.x86_64.a scene/libscene.osx.opt.tools.x86_64.a servers/libservers.osx.opt.tools.x86_64.a core/libcore.osx.opt.tools.x86_64.a modules/freetype/libfreetype_builtin.osx.opt.tools.x86_64.a -lpthread -liconv -lpthread
Undefined symbols for architecture x86_64:
  "___isPlatformVersionAtLeast", referenced from:
      _free_chunklist in libmonosgen-2.0.a(libmonoutils_la-mono-codeman.o)
      _new_codechunk in libmonosgen-2.0.a(libmonoutils_la-mono-codeman.o)
      _mono_codeman_enable_write in libmonosgen-2.0.a(libmonoutils_la-mono-codeman.o)
      _mono_codeman_disable_write in libmonosgen-2.0.a(libmonoutils_la-mono-codeman.o)
      _codechunk_valloc in libmonosgen-2.0.a(libmonoutils_la-mono-codeman.o)
ld: symbol(s) not found for architecture x86_64
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
scons: *** [bin/godot.osx.opt.tools.x86_64.mono] Error 1
scons: building terminated because of errors.
```

Here's a more minimal example:
```c
#include <stdio.h>
int main() {
  if (__builtin_available(macOS 11, *)) {
    printf("macOS 11 available\n");
  }
  return 0;
}
```

With osxcross:
```
$ x86_64-apple-darwin20.2-clang -mmacosx-version-min=10.12 main.c
Undefined symbols for architecture x86_64:
  "___isPlatformVersionAtLeast", referenced from:
      _main in main-5a94c9.o
ld: symbol(s) not found for architecture x86_64
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)

$ x86_64-apple-darwin20.2-clang -mmacosx-version-min=11.0 main.c
// success
```

With Apple Clang on a Mac running Big Sur:
```
$ clang -mmacosx-version-min=10.12 main.c
$ ./a.out 
macOS 11 available

$ clang -mmacosx-version-min=11.0 main.c
$ ./a.out 
macOS 11 available
```

<details>
<summary>Side note on checking against future SDKs</summary>

As a side note, Apple Clang also seems to allow checking for theoretical future SDK versions:

```c
#include <stdio.h>
int main() {
  if (__builtin_available(macOS 11, *)) {
    printf("macOS 11 available\n");
  }
  if (__builtin_available(macOS 12, *)) {
    printf("macOS 12 available\n");
  }
  return 0;
}
```

```
$ clang -mmacosx-version-min=11.0 main.c
$ ./a.out 
macOS 11 available

$ clang -mmacosx-version-min=12.0 main.c
$ ./a.out 
macOS 11 available
macOS 12 available
```

Where osxcross errors:
```
$ x86_64-apple-darwin20.2-clang -mmacosx-version-min=12.0 main.c
osxcross: error: targeted macOS version must be <= 11.1.0 (SDK)
osxcross: error: while detecting target
```

But that's not a big deal compared to the main linking issue reported above. And I'm not sure Apple Clang's behavior even makes sense since it lets me run a binary on Big Sur compiled with `mmacosx-version-min=12.0`...

</details>

--- a/README.COMPILER-RT.md
+++ b/README.COMPILER-RT.md
@@ -10,6 +10,10 @@ Ensure you have finished `build.sh`,
 
 then run: `./build_compiler_rt.sh`.
 
+By default, installation steps for compiler-rt will be printed to the terminal
+to run manually, but you can automate the installation process by defining
+`ENABLE_COMPILER_RT_INSTALL`.
+
 You can verify compiler-rt is working by invoking the following command:
 
     echo "int main(void){return 0;}" | xcrun clang -xc -o/dev/null -v - 2>&1 | \

--- a/README.md
+++ b/README.md
@@ -33,6 +33,9 @@ It also includes scripts for optionally building
 * the "compiler-rt" runtime library, and
 * the `llvm-dsymutil` tool required for debugging.
 
+Note: The "compiler-rt" library can be needed to link code that uses the
+`__builtin_available()` runtime version check.
+
 
 ### WHAT CAN BE BUILT WITH IT? ###
 

--- a/build_compiler_rt.sh
+++ b/build_compiler_rt.sh
@@ -182,22 +182,39 @@ fi
 rm -f $BUILD_DIR/.compiler-rt_build_complete
 
 
+# Installation. Can be either automated (ENABLE_COMPILER_RT_INSTALL) or will
+# print the commands that the user should run manually.
+
+function print_or_run() {
+  if [ -z "$ENABLE_COMPILER_RT_INSTALL" ]; then
+    echo "$@"
+  else
+    $@
+  fi
+}
+
 echo ""
 echo ""
 echo ""
-echo "Please run the following commands by hand to install compiler-rt:"
+if [ -z "$ENABLE_COMPILER_RT_INSTALL" ]; then
+  echo "Please run the following commands by hand to install compiler-rt:"
+else
+  echo "Installing compiler-rt headers and libraries to the following paths:"
+  echo "  ${CLANG_INCLUDE_DIR}"
+  echo "  ${CLANG_DARWIN_LIB_DIR}"
+fi
 echo ""
 
-echo "mkdir -p ${CLANG_INCLUDE_DIR}"
-echo "mkdir -p ${CLANG_DARWIN_LIB_DIR}"
-echo "cp -rv $BUILD_DIR/compiler-rt/compiler-rt/include/sanitizer ${CLANG_INCLUDE_DIR}"
+print_or_run mkdir -p ${CLANG_INCLUDE_DIR}
+print_or_run mkdir -p ${CLANG_DARWIN_LIB_DIR}
+print_or_run cp -rv $BUILD_DIR/compiler-rt/compiler-rt/include/sanitizer ${CLANG_INCLUDE_DIR}
 
 if [ $USE_CMAKE -eq 1 ]; then
 
   ### CMAKE ###
 
-  echo "cp -v $BUILD_DIR/compiler-rt/compiler-rt/build/lib/darwin/*.a ${CLANG_DARWIN_LIB_DIR}"
-  echo "cp -v $BUILD_DIR/compiler-rt/compiler-rt/build/lib/darwin/*.dylib ${CLANG_DARWIN_LIB_DIR}"
+  print_or_run cp -v $BUILD_DIR/compiler-rt/compiler-rt/build/lib/darwin/*.a ${CLANG_DARWIN_LIB_DIR}
+  print_or_run cp -v $BUILD_DIR/compiler-rt/compiler-rt/build/lib/darwin/*.dylib ${CLANG_DARWIN_LIB_DIR}
 
   ### CMAKE END ###
 
@@ -209,7 +226,7 @@ else
 
   function print_install_command() {
     if [ -f "$1" ]; then
-      echo "cp $PWD/compiler-rt/$1 ${CLANG_DARWIN_LIB_DIR}/$2"
+      print_or_run cp $PWD/compiler-rt/$1 ${CLANG_DARWIN_LIB_DIR}/$2
     fi
   }
 
@@ -219,13 +236,11 @@ else
   print_install_command "cc_kext/libcompiler_rt.a"     "libclang_rt.cc_kext.a"
   print_install_command "profile_osx/libcompiler_rt.a" "libclang_rt.profile_osx.a"
 
-
   print_install_command "ubsan_osx_dynamic/libcompiler_rt.dylib" \
     "libclang_rt.ubsan_osx_dynamic.dylib"
 
   print_install_command "asan_osx_dynamic/libcompiler_rt.dylib" \
     "libclang_rt.asan_osx_dynamic.dylib"
-
 
   popd &>/dev/null
 

